File: bp++/binpac/pactypes/iterator.py
# $Id$
#
# Generic iterator operators.

import binpac.type as type
import binpac.operator as operator

@operator.Deref(type.Iterator)
class _:
    def type(iter):
        return iter.type().derefType()

    def evaluate(cg, iter):
        tmp = cg.functionBuilder().addLocal("__elem", iter.type().derefType().hiltiType(cg))
        cg.builder().deref(tmp, iter.evaluate(cg))
        return tmp

@operator.IncrPrefix(type.Iterator)
class _:
    def type(iter):
        return iter.type()

    def evaluate(cg, iter):
        op = iter.evaluate(cg)
        cg.builder().incr(op, op)
        return op

@operator.IncrPostfix(type.Iterator)
class _:
    def type(iter):
        return iter.type()

    def evaluate(cg, iter):
        tmp = cg.functionBuilder().addLocal("__tmp", iter.type().hiltiType(cg))
        op = iter.evaluate(cg)
        cg.builder().assign(tmp, op)
        cg.builder().incr(op, op)
        return tmp

@operator.PlusAssign(type.Iterator, type.UnsignedInteger)
class _:
    def type(iter, incr):
        return iter.type()

    def evaluate(cg, iter, incr):
        op1 = iter.evaluate(cg)
        op2 = incr.evaluate(cg)
        cg.builder().incr_by(op1, op1, op2)
        return op1

@operator.Plus(type.Iterator, type.UnsignedInteger)
class _:
    def type(iter, incr):
        return iter.type()

    def evaluate(cg, iter, incr):
        op1 = iter.evaluate(cg)
        op2 = incr.coerceTo(type.UnsignedInteger(64), cg).evaluate(cg)
        tmp = cg.functionBuilder().addLocal("__niter", op1.type())
        cg.builder().incr_by(tmp, op1, op2)
        return tmp

@operator.Equal(type.Iterator, type.Iterator)
class _:
    def type(iter1, iter2):
        return type.Bool()

    def evaluate(cg, iter1, iter2):
        tmp = cg.functionBuilder().addLocal("__equal", hilti.type.Bool())
        cg.builder().equal(tmp, iter1.evaluate(cg), iter2.evaluate(cg))
        return tmp

File: spicy/codegen/operators/iterator.cc

#include "cg-operator-common.h"
#include <spicy/autogen/operators/iterator.h>

using namespace spicy;
using namespace spicy::codegen;

void CodeBuilder::visit(expression::operator_::iterator::Equal* i)
{
    auto result = builder()->addTmp("equal", hilti::builder::boolean::type());
    auto op1 = cg()->hiltiExpression(i->op1());
    auto op2 = cg()->hiltiExpression(i->op2());
    cg()->builder()->addInstruction(result, hilti::instruction::operator_::Equal, op1, op2);
    setResult(result);
}

void CodeBuilder::visit(expression::operator_::iterator::Deref* i)
{
    auto itype = ast::rtti::checkedCast<type::Iterator>(i->op1()->type());
    auto etype = cg()->hiltiType(
        ast::type::checkedTrait<type::trait::Iterable>(itype->argType())->elementType());

    auto result = builder()->addTmp("elem", etype);
    auto op1 = cg()->hiltiExpression(i->op1());
    cg()->builder()->addInstruction(result, hilti::instruction::operator_::Deref, op1);
    setResult(result);
}

void CodeBuilder::visit(expression::operator_::iterator::IncrPostfix* i)
{
    auto itype = cg()->hiltiType(i->op1()->type());
    auto result = builder()->addTmp("iter", itype);
    auto op1 = cg()->hiltiExpression(i->op1());
    cg()->builder()->addInstruction(result, hilti::instruction::operator_::Assign, op1);
    cg()->builder()->addInstruction(op1, hilti::instruction::operator_::Incr, op1);
    setResult(result);
}

void CodeBuilder::visit(expression::operator_::iterator::IncrPrefix* i)
{
    auto itype = cg()->hiltiType(i->op1()->type());
    auto op1 = cg()->hiltiExpression(i->op1());
    cg()->builder()->addInstruction(op1, hilti::instruction::operator_::Incr, op1);
    setResult(op1);
}

// We don't have a generic incrby hilti operator yet.

void CodeBuilder::visit(expression::operator_::iterator::Plus* i)
{
    auto itype = cg()->hiltiType(i->op1()->type());
    auto result = builder()->addTmp("iter", itype);
    auto op1 = cg()->hiltiExpression(i->op1());
    auto op2 = cg()->hiltiExpression(i->op2(), std::make_shared<type::Integer>(64, false));
    cg()->builder()->addInstruction(result, hilti::instruction::operator_::Assign, op1);
    cg()->builder()->addInstruction(result, hilti::instruction::operator_::IncrBy, result, op2);
    setResult(result);
}

void CodeBuilder::visit(expression::operator_::iterator::PlusAssign* i)
{
    auto op1 = cg()->hiltiExpression(i->op1());
    auto op2 = cg()->hiltiExpression(i->op2());

    cg()->builder()->addInstruction(op1, hilti::instruction::operator_::IncrBy, op1, op2);
    setResult(op1);
}
